File: packages/flet/lib/src/utils/box.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';

BoxConstraints? parseBoxConstraints(dynamic value,
    [BoxConstraints? defaultValue]) {
  if (value == null) return defaultValue;

  return BoxConstraints(
    minHeight: parseDouble(value["min_height"], 0.0)!,
    minWidth: parseDouble(value["min_width"], 0.0)!,
    maxHeight: parseDouble(value["max_height"], double.infinity)!,
    maxWidth: parseDouble(value["max_width"], double.infinity)!,
  );
}

List<BoxShadow>? parseBoxShadows(dynamic value, ThemeData theme,
    [List<BoxShadow>? defaultValue]) {
  if (value == null) return defaultValue;
  if (value is List) {
    return value.map((e) => parseBoxShadow(e, theme)!).toList();
  } else {
    return [parseBoxShadow(value, theme)!];
  }
}

BoxShadow? parseBoxShadow(dynamic value, ThemeData theme,
    [BoxShadow? defaultValue]) {
  if (value == null) return defaultValue;

  return BoxShadow(
      color: parseColor(value["color"], theme, Colors.black)!,
      offset: parseOffset(value["offset"], Offset.zero)!,
      blurStyle: parseBlurStyle(value["blur_style"], BlurStyle.normal)!,
      blurRadius: parseDouble(value["blur_radius"], 0)!,
      spreadRadius: parseDouble(value["spread_radius"], 0)!);
}

BoxDecoration? parseBoxDecoration(dynamic value, BuildContext context,
    [BoxDecoration? defaultValue]) {
  if (value == null) return defaultValue;
  var theme = Theme.of(context);

  var shape = parseBoxShape(value["shape"], BoxShape.rectangle)!;
  var borderRadius = parseBorderRadius(value["border_radius"]);
  var color = parseColor(value["bgcolor"], theme);
  var gradient = parseGradient(value["gradient"], theme);
  var blendMode = parseBlendMode(value["blend_mode"]);

  return BoxDecoration(
    color: color,
    border: parseBorder(value["border"], theme),
    shape: shape,
    borderRadius: shape == BoxShape.circle ? null : borderRadius,
    backgroundBlendMode: color != null || gradient != null ? blendMode : null,
    boxShadow: parseBoxShadows(value["shadows"], theme),
    gradient: gradient,
    image: parseDecorationImage(value["image"], context),
  );
}

BoxDecoration? boxDecorationFromDetails({
  Color? color,
  Border? border,
  BoxShape? shape,
  BorderRadius? borderRadius,
  BlendMode? blendMode,
  List<BoxShadow>? boxShadow,
  Gradient? gradient,
  DecorationImage? image,
}) {
  // If no custom properties are provided, return null
  if (!(color != null ||
      border != null ||
      borderRadius != null ||
      gradient != null ||
      shape != null ||
      boxShadow != null ||
      image != null)) {
    return null;
  }

  return BoxDecoration(
    color: color,
    border: border,
    backgroundBlendMode: color != null || gradient != null ? blendMode : null,
    borderRadius: shape == BoxShape.circle ? null : borderRadius,
    gradient: gradient,
    shape: shape ?? BoxShape.rectangle,
    boxShadow: boxShadow,
    image: image,
  );
}

DecorationImage? parseDecorationImage(dynamic value, BuildContext context,
    [DecorationImage? defaultValue]) {
  if (value == null) return defaultValue;

  var src = value["src"];
  ImageProvider? image = parseImageProvider(src, context);
  if (image == null) return defaultValue;

  return DecorationImage(
    image: image,
    colorFilter: parseColorFilter(value["color_filter"], Theme.of(context)),
    fit: parseBoxFit(value["fit"]),
    alignment: parseAlignment(value["alignment"], Alignment.center)!,
    repeat: parseImageRepeat(value["repeat"], ImageRepeat.noRepeat)!,
    matchTextDirection: parseBool(value["match_text_direction"], false)!,
    scale: parseDouble(value["scale"], 1.0)!,
    opacity: parseDouble(value["opacity"], 1.0)!,
    filterQuality:
        parseFilterQuality(value["filter_quality"], FilterQuality.medium)!,
    invertColors: parseBool(value["invert_colors"], false)!,
    isAntiAlias: parseBool(value["anti_alias"], false)!,
  );
}

extension BoxParsers on Control {
  BoxConstraints? getBoxConstraints(String propertyName,
      [BoxConstraints? defaultValue]) {
    return parseBoxConstraints(get(propertyName), defaultValue);
  }

  List<BoxShadow>? getBoxShadows(String propertyName, ThemeData theme,
      [List<BoxShadow>? defaultValue]) {
    return parseBoxShadows(get(propertyName), theme, defaultValue);
  }

  BoxDecoration? getBoxDecoration(String propertyName, BuildContext context,
      [BoxDecoration? defaultValue]) {
    return parseBoxDecoration(get(propertyName), context, defaultValue);
  }

  DecorationImage? getDecorationImage(String propertyName, BuildContext context,
      [DecorationImage? defaultValue]) {
    return parseDecorationImage(get(propertyName), context, defaultValue);
  }
}

File: sdk/python/packages/flet-core/src/flet_core/box.py
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional, List, Union

from flet_core.alignment import Alignment
from flet_core.border import Border
from flet_core.gradients import Gradient
from flet_core.types import (
    OffsetValue,
    BorderRadiusValue,
    BlendMode,
    ImageFit,
    OptionalNumber,
    ImageRepeat,
)


@dataclass
class ColorFilter:
    color: Optional[str] = field(default=None)
    blend_mode: Optional[BlendMode] = field(default=None)


class FilterQuality(Enum):
    NONE = "none"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class ShadowBlurStyle(Enum):
    NORMAL = "normal"
    SOLID = "solid"
    OUTER = "outer"
    INNER = "inner"


@dataclass
class BoxShadow:
    spread_radius: Optional[float] = field(default=None)
    blur_radius: Optional[float] = field(default=None)
    color: Optional[str] = field(default=None)
    offset: OffsetValue = field(default=None)
    blur_style: ShadowBlurStyle = field(default=ShadowBlurStyle.NORMAL)


class BoxShape(Enum):
    RECTANGLE = "rectangle"
    CIRCLE = "circle"


@dataclass
class DecorationImage:
    src: Optional[str] = None
    src_base64: Optional[str] = None
    color_filter: Optional[ColorFilter] = None
    fit: Optional[ImageFit] = None
    alignment: Optional[Alignment] = None
    repeat: Optional[ImageRepeat] = None
    match_text_direction: Optional[bool] = None
    scale: OptionalNumber = None
    opacity: OptionalNumber = None
    filter_quality: Optional[FilterQuality] = None
    invert_colors: Optional[bool] = None
    anti_alias: Optional[bool] = None


@dataclass
class BoxDecoration:
    bgcolor: Optional[str] = None
    image: Optional[DecorationImage] = None
    border: Optional[Border] = None
    border_radius: BorderRadiusValue = None
    shadow: Union[None, BoxShadow, List[BoxShadow]] = None
    gradient: Optional[Gradient] = None
    shape: Optional[BoxShape] = None
    blend_mode: Optional[BlendMode] = None
